Python/ceval.c

```diff
@@ -2163,9 +2163,18 @@ _PyEval_UnpackIterableStackRef(PyThreadState *tstate, _PyStackRef v_stackref,
             return 1;
         }
         Py_DECREF(w);
-        _PyErr_Format(tstate, PyExc_ValueError,
-                      "too many values to unpack (expected %d)",
-                      argcnt);
+
+        ll = PyObject_Size(v);
+        if (ll < 0) {
```

Comment:
```suggestion
        if (ll <= argcnt) {
```

Please consider, and test, evil classes that lie about their size. For example:
```py
class EvilLiar:
  def __len__(self):
      return 2

  def __iter__(self):
      yield from (1, 2, 3)

a, b = EvilLiar()
```

At this level, I don't think we want to print `(expected 2, got 2)`, even if user is at fault here.

In other evil classes, `__len__` can fail with an exception. Please test this.
IMO, if it's not a `TypeError`/`AttributeError`, and *especially* if it's a `BaseException` but not `Exception`, it needs to be chained (e.g. using `PyException_SetContext`).

Other evil classes can have side effects, or take a long time. IMO, that's OK -- an extra `__len__` call *shouldn't* hurt.

Reply:
Thanks for catching this, I'll test and take care of these cases.

Reply:
I think I've handled the weird cases now, but I couldn't figure out what would raise an `AttributeError` in the normal case, so I'm only ignoring `TypeError`'s context right now.

There's also a bunch of code duplication for `_PyErr_Format`, should it be deduped in some way?